plugins/wasmembed/abi.cpp

```diff
@@ -124,12 +129,13 @@ T load_int(const wasmtime::Span<uint8_t> &data, int32_t ptr)
     auto nbytes = sizeof(retVal);
     for (int i = 0; i < nbytes; ++i)
     {
-        uint8_t b = data[ptr + i];
+        std::make_unsigned_t<T> b = data[ptr + i];
         retVal += b << (i * 8);
     }
-    if (std::is_signed<T>::value)
+    if (std::is_signed<T>::value && data[ptr + nbytes - 1] & 0x80)
```

Comment:
There also needs to be a check that nbytes != sizeof(T), because value << 8 * sizeof(T) is undefined behaviour.

There should also be a comment
//if the top bit is set then sign extend the result by setting the top bits.

NOTE: Running in debug on linux will throw errors when undefined behaviour is spotted.